pepper-apis/src/main/java/org/broadinstitute/ddp/service/PdfGenerationService.java

```diff
@@ -138,49 +139,42 @@ private static void setIsChecked(PdfFormField field, boolean isChecked) {
-                            pdfOrderIndex, configuration.getConfigName(), participant, instances, errors);
-                    break;
-                case MAILING_ADDRESS:
-                    pdf = generateMailingAddressPdf((MailingAddressTemplate) template, participant.getUser(), errors,
-                            configuration.getStudyGuid(), handle);
-                    break;
-                default:
-                    errors.add("Tried to use a template type (" + template.getType() + ") that is unsupported.");
-                    pdf = null;
-                    break;
-            }
-            if (pdf != null && pdf.length != 0) {
-                individualPdfs.add(pdf);
-            }
-        }
-        checkForErrors(errors, userGuid);
-
+        byte[] pdf;
         try (
                 ByteArrayOutputStream renderedStream = new ByteArrayOutputStream();
                 PdfWriter pdfWriter = new PdfWriter(renderedStream);
                 PdfDocument mergedDoc = new PdfDocument(pdfWriter)) {
-            // concatenate each rendered doc to a master doc
-
             int counter = 0;
-            for (byte[] pdfStream : individualPdfs) {
-                copyPdfToMasterDoc(pdfStream, counter, mergedDoc);
+            for (int pdfOrderIndex = 0; pdfOrderIndex < configuration.getTemplateIds().size(); pdfOrderIndex++) {
+                Long templateId = configuration.getTemplateIds().get(pdfOrderIndex);
+                PdfTemplate template = handle.attach(PdfDao.class)
+                        .findFullTemplateByTemplateId(templateId)
+                        .orElseThrow(() -> new DDPException("Could not find template with id:" + templateId));
+                switch (template.getType()) {
+                    case PHYSICIAN_INSTITUTION:
+                        pdf = generatePhysicianInstitutionPdf(configuration.getStudyGuid(),
+                                (PhysicianInstitutionTemplate) template, pdfOrderIndex, configuration.getConfigName(), participant, errors);
+                        break;
+                    case CUSTOM:
+                        pdf = generateCustomPdf((CustomTemplate) template,
+                                pdfOrderIndex, configuration.getConfigName(), participant, instances, errors);
+                        break;
+                    case MAILING_ADDRESS:
+                        pdf = generateMailingAddressPdf((MailingAddressTemplate) template, participant.getUser(), errors,
+                                configuration.getStudyGuid(), handle);
+                        break;
+                    default:
+                        errors.add("Tried to use a template type (" + template.getType() + ") that is unsupported.");
+                        pdf = null;
+                        break;
+                }
+                if (pdf != null && pdf.length != 0) {
+                    copyPdfToMasterDoc(pdf, counter, mergedDoc);
+                }
```

Comment:
Do you think we should add a `pdf = null` statement here? So that we "drop" that reference at the end of this loop iteration, so that in the next loop iteration, if the JVM needs to run GC before loading in the result of `findFullTemplateByTemplateId()` then it can free up the `byte[]`? We could also simply just move the `byte[] pdf` variable declaration inside the `for` loop. That should drop that byte array reference after each iteration. Or am I totally misunderstanding things here?

Reply:
Nulling it variable explicitly ought to be unnecessary. Assigning a new object to that variable and stack frame for this method getting unloaded both ought to de-reference the PDF object and make it candidate for GC reaping.
But moving the variable into the loop is a good idea. Probably marginal benefit, but definitely in the spirit of what I am trying to do in PR.
7137e2eb